src/components/TypeAhead.tsx

```diff
@@ -102,4 +110,61 @@ class TypeAhead extends React.PureComponent<TypeaheadProps, TypeaheadState> {
   }
 }
 
-export default TypeAhead;
+export class CachedTypeAhead extends React.PureComponent<
+  CachedTypeAheadProps,
+  CacheTypeAheadState
+> {
+  state = {
+    query: "",
+    cache: {
+      "": {
+        promise: Promise.resolve([]),
+        options: []
+      }
+    }
+  };
+
+  onSearchBoxQueryChange = async (query: string) => {
+    if (this.state.query === query) return;
+
+    if (this.state.cache[query]) {
+      return this.setState({ query });
+    }
+
+    const optionsPromise = this.props.children(query);
+    this.setState({
+      query: query,
+      cache: {
+        ...this.state.cache,
+        [query]: {
+          promise: optionsPromise
```

Comment:
why save promise when it is never used from state?

Reply:
also currently there is a flicker when new data is fetched. the drop-down clears and then when data is fetched, it comes back as for some period of time there are no results. I don't think that is the expected behavior?

Reply:
Was initially hoping to cancel the Promise on character change, but realized that it's difficult to expose cancellation token while keeping functionality simple (i.e. children function returns Promise to `<Option />`s)
Need to do some more research before finalize API.